File: cli/src/specsfy_cli/backlog.py
from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass
from pathlib import Path


FIELD = re.compile(r"^\*\*(?P<name>[^*]+)\*\*:\s*(?P<value>.+?)\s*$")
TABLE_FIELD = re.compile(
    r"^\|\s*(?P<name>[^|]+?)\s*\|\s*(?P<value>[^|]+?)\s*\|\s*$"
)


@dataclass(frozen=True)
class BacklogItem:
    slug: str
    title: str
    identifier: str
    status: str
    path: Path
    content: str


def scan_backlogs(project: Path) -> list[BacklogItem]:
    root = project.expanduser().resolve()
    return [_parse_backlog(path) for path in _backlog_paths(root)]


def backlogs_fingerprint(project: Path) -> str:
    root = project.expanduser().resolve()
    digest = hashlib.sha256()
    for path in _backlog_paths(root):
        digest.update(str(path.relative_to(root)).encode())
        digest.update(b"\0")
        try:
            digest.update(path.read_bytes())
        except OSError:
            continue
        digest.update(b"\0")
    return digest.hexdigest()


def _backlog_paths(root: Path) -> list[Path]:
    return sorted(
        path for path in root.glob("specs/backlog/*.md") if path.is_file()
    )


def _parse_backlog(path: Path) -> BacklogItem:
    content = path.read_text(encoding="utf-8")
    title = path.stem
    fields: dict[str, str] = {}
    for line in content.splitlines():
        if line.startswith("# ") and title == path.stem:
            title = line.removeprefix("# ").strip()
            if title.lower().startswith("backlog:"):
                title = title.split(":", 1)[1].strip()
        field = FIELD.match(line) or TABLE_FIELD.match(line)
        if field:
            fields[field.group("name").strip().lower()] = field.group("value").strip()
    return BacklogItem(
        slug=path.stem,
        title=title,
        identifier=fields.get("id", path.stem),
        status=fields.get("status", "Unknown"),
        path=path,
        content=content,
    )

File: cli/src/backlog.ts
/**
 * Leitura dos itens de backlog de um projeto consumidor.
 */

import { createHash } from "node:crypto";
import { readFile, readdir } from "node:fs/promises";
import { basename, join, relative } from "node:path";
import { isFile, resolvePath } from "./filesystem.js";

const FIELD = /^\*\*([^*]+)\*\*:\s*(.+?)\s*$/;
const TABLE_FIELD = /^\|\s*([^|]+?)\s*\|\s*([^|]+?)\s*\|\s*$/;

/** Projeção de um arquivo em `specs/backlog/`. */
export interface BacklogItem {
  slug: string;
  title: string;
  identifier: string;
  status: string;
  path: string;
  content: string;
}

/** Lê todos os backlogs do projeto em ordem estável. */
export async function scanBacklogs(project: string): Promise<BacklogItem[]> {
  const root = resolvePath(project);
  return Promise.all((await backlogPaths(root)).map(parseBacklog));
}

/** Calcula o fingerprint dos caminhos e conteúdos de backlog. */
export async function backlogsFingerprint(project: string): Promise<string> {
  const root = resolvePath(project);
  const digest = createHash("sha256");
  for (const path of await backlogPaths(root)) {
    digest.update(relative(root, path));
    digest.update("\0");
    try {
      digest.update(await readFile(path));
    } catch {
      continue;
    }
    digest.update("\0");
  }
  return digest.digest("hex");
}

async function backlogPaths(root: string): Promise<string[]> {
  const directory = join(root, "specs", "backlog");
  let names: string[];
  try {
    names = await readdir(directory);
  } catch {
    return [];
  }
  const paths = names
    .filter((name) => name.endsWith(".md"))
    .map((name) => join(directory, name));
  const checks = await Promise.all(paths.map(isFile));
  return paths.filter((_, index) => checks[index]).sort();
}

async function parseBacklog(path: string): Promise<BacklogItem> {
  const content = await readFile(path, "utf8");
  const slug = basename(path, ".md");
  let title = slug;
  const fields = new Map<string, string>();
  for (const line of content.split(/\r?\n/u)) {
    if (line.startsWith("# ") && title === slug) {
      title = line.slice(2).trim();
      if (title.toLowerCase().startsWith("backlog:")) {
        title = title.split(":", 2)[1]?.trim() ?? title;
      }
    }
    const match = FIELD.exec(line) ?? TABLE_FIELD.exec(line);
    if (match?.[1] && match[2]) {
      fields.set(match[1].trim().toLowerCase(), match[2].trim());
    }
  }
  return {
    slug,
    title,
    identifier: fields.get("id") ?? slug,
    status: fields.get("status") ?? "Unknown",
    path,
    content,
  };
}
